### ISSUE TITLE
"Set brightness" doesn't work as expected on Chromebooks

### ISSUE THREAD

**Issue by jwzimmer** [LLK/scratch-flash#1312]
**TL;DR:** @seotts ran into an issue that seems to be related to the issue described in https://github.com/LLK/scratch-flash/issues/570, as well as potentially being caused by (some?) Chromebooks not working properly with Stage3D. I'm filing it as separate from #570 in case there is an additional issue going on with Chromebooks and Stage3D.

If we want to look into this further, we should determine if there is an issue with Chromebooks + Stage3D, if so, which ones, and if not, this may have the same root cause as #570 and then could be closed as a dupe.

> ... while people are following along the dance cards, at some point "set brightness" is behaving like "set ghost effect" and a sprite is hiding
> they're working on chrome books, and it's only happening sometimes
> the dance cards involve setting brightness to -100

Note that if you follow the dance cards exactly, you can encounter this bug - at least two of the group @seotts was working with did.

@seotts found from https://wiki.scratch.mit.edu/wiki/Stage3D that using ctrl-m fixes the bug.
> i think chromebooks (or at least the ones we have over here) don't work with the Stage3D or something

**Comment by cwillisf** [comment]
Interesting! I wonder if this is a shader bug that most drivers / hardware can work around, or if it's a bug in the drivers or hardware used by those Chromebooks...

@seotts: if you still have access to the Chromebooks in question, could you try the brightness block on them in [Scratch 3.0](https://llk.github.io/scratch-gui/) and see if there's a similar bug there? Thanks!

**Comment by rschamp** [comment]
@cwillisf @thisandagain Would this be prio-critical if true? If not, should we try to repro?

**Comment by thisandagain** [comment]
Depending on the repro steps / conditions I think this could end up being `prio-critical`. Because of this I'd like to see us attempt to repro.

**Comment by TheLogFather** [comment]
A few times I've heard mention that brightness effect appears to be different on different machines (also chromebooks, IIRC).
In particular, the way it deals with black. I think they were turning black into white, whereas most leave it black. (Here's an example: https://scratch.mit.edu/projects/94409099/#comments-54376684)

I'd be curious to know if there are differences in the way that https://github.com/LLK/scratch-flash/pull/972 behaves on chromebooks...

**Comment by jwzimmer** [comment]
### Attempts to repro

- While testing, I noticed that the performance is very weird. I took a video of it.

**Comment by cwillisf** [comment]
The performance is definitely strange... it kind of looks like a caching thing but I can't currently explain why the slowness would last as long as it does.

For example, the first time a particular combination of effects is used, the renderer has to build a new shader to handle that combination. If that combination is used again then the renderer can reuse the cached shader. It looks like this project only uses maybe 2 combinations of effects, though: no effects, and just color effect. I would expect that to mean two slow frames -- not 15 seconds worth of slow frames.

**Comment by jwzimmer** [comment]
I talked about this with @cwillisf - even though I didn't repro the issue as @seotts described it, I did repro the brightness not behaving the same way on Mac & Chromebook, and @cwillisf thought it was pretty likely that these issues have the same root (if, after investigation, it turns out they don't, we can always split this into two github issues):
> it's probably something that our shader code is doing which isn't quite compatible with whatever graphics chip is in those Chromebooks... they're probably two manifestations of the same bug, but if I'm wrong then they could be separate bugs

**Comment by rschamp** [comment]
If we get reports from more users experiencing this we will revisit.

**Comment by chrisgarrity** [comment]
I had one of my sons check on a Chromebook pixel - it worked fine in scratch 2.0 (online) and the new prototype.

**Comment by jwzimmer** [comment]
Thanks @chrisgarrity!
